<center> <h2> DS 3000 - Fall 2020</h2> </center>
<center> <h3> DS Report </h3> </center>


<center> <h3> Predicting the Price for a Ride-Share Service</h3> </center>
<center><h4>Aneesh Atri, Siddhant Dosi, Justin Lau</h4></center>


<hr style="height:2px; border:none; color:black; background-color:black;">

#### Executive Summary:

Our goal in this project was to develop an algorithm that can predict the price of a ride-share ride in Boston based on the location, distance, time, cab type and weather related information. We first made use of visualizations and hypothesis tests to highlight the effect that certain variables (like time of day and cab type) have on the price of the ride. We then cleaned our data set by removing redundant or invalid features, scaling our data, using one-hot encoding to convert categorical data into quantitative data, separating data into a testing and training set, and removing features through model-based feature selection. We finally used three machine learning algorithms: Multiple Linear Regression, Ridge Regression, and Lasso Regression to create a model that can predict the price of the ride based on the available features. The best performing model was the Ridge Regression model, which had an R^2 value on the testing set of 0.92805 and MSE of 6.264, suggesting that our model can account for nearly 93% of the variation in price of ride-share rides in Boston.


<hr style="height:2px; border:none; color:black; background-color:black;">

## Outline
1. <a href='#1'>INTRODUCTION</a>
2. <a href='#2'>METHOD</a>
3. <a href='#3'>RESULTS</a>
4. <a href='#4'>DISCUSSION</a>

<a id="1"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 1. INTRODUCTION

The purpose of our project is to predict the price of ridesharing services through Uber and Lyft in Boston based on features like the distance of the ride, weather and time of day. These apps require users to put in their current location and their desired destination, which would seem to imply that distance is the only important factor in calculating the cost of a ride. However, our group believes that there are more factors that go into play when the price of a ride is calculated. We want to see if with additional features like weather and time, we can create a machine learning model to more accurately predict the price of an Uber or Lyft. In essence, our goal is to learn what factors are considered when these apps calculate the price of a ride using their service. We will compare three different supervised machine learning regression algorithms, including Multiple Linear Regression, Ridge Regression, and Lasso Regression models.

We believe that this is an important problem to tackle as it can help users plan their rides in advance. An application of a price predictor model would be to create an application that can allow users to input information about their ride (the time, starting and ending locations, weather, and cab type), and give them information about how much the ride would be expected to cost. This can allow potential users to plan how much they would need to spend days in advance. The model would also provide insights related to other factors, such as whether Lyft or Uber would be the better option, hours of the day to avoid rides, and which locations may be more expensive than others. 

There has been significant prior work on this topic. Govind Pramod Yatnalkar of Marshall University used data on Uber users to classify them into one of 5 categories for what they prefer most in a driver: chattiness, safety, punctuality, friendliness, and comfortability. Yatnalkar used a Support Vector Machine classification Machine Learning algorithm and came up with excellent results. He had an overall accuracy of 91.65%, a root Mean Squared Error of 0.64, and an F1 score of roughly 91.5% on average (Yatnalkar, 2019). Meanwhile, Dawn Woodard, Senior Data Science Manager of Dynamic Pricing at Uber, confirms that Uber uses algorithms that increase the price of rides when demand for rides outpaces the supply of drivers, and lowers the price when the opposite is true (Woodard, 2018). This confirms our suspicion that the time of day could greatly affect the price of a ride-share ride. However, the most similar project to ours was completed by Malaiiyyappan Muthhukumar in 2019, who used the same Boston Uber and Lyft dataset to predict the price of a ride. She used Linear Regression, Elastic Net, XGB Regression, and Random Forest Regression. Her best performance was from her XGB Regression Model, which had an R^2 of nearly 0.96 on the testing set, and a RMSE of 2 (Muthhukumar, 2019). Despite the similarities of these projects, we seek to complete this project to corroborate the results that Muthhukumar received, and also use other algorithms to see if we can improve on her performance.

Based on the problem that we are looking to tackle, there are many questions that we seek to answer related to our project. These include:
* Which features contribute the most to the price of a ride-share ride?
* Which individual feature has the greatest impact on the price of a ride-share ride?
* Are there any features in the dataset that have no/minimal impact on the price?
* Is there a difference in price between Uber and Lyft rides (holding all other factors constant)?
* What percentage of the variation in price can be explained by a regression model based on the features of the data set?
* Would removing some features through univariate or model-based feature selection improve the strength of a regression model?
* Is a multi-linear regression model the best for this dataset, or is it better to tune the model with a Ridge or Lasso regression model?

There are also certain hypotheses that we would like to test to determine which variables have an impact on the price of the ride. One such test is a difference of means test between the mean price of a Lyft and Uber ride. Another test is a difference of means test between the mean price of a ride at a “rush hour” vs a ride at a “non rush hour”.

Note: Sources are cited at the end of this report

<a id="2"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 2. METHOD

### 2.1. Data Acquisition

The data for this project was obtained from Kaggle, the data science community. Here is the link to the original dataset: https://www.kaggle.com/brllrb/uber-and-lyft-dataset-boston-ma. As a group, we wanted an easier way to access the data, so we uploaded the dataset to Dropbox. The dataset can be accessed using this link: https://www.dropbox.com/s/1nyqv980whwhn2q/rideshare_kaggle.csv?dl=1. We also added other components like the graphs to a github repository: https://github.com/aneesh1/ds3000project. 
Our dataset contains information about Uber and Lyft rides in Boston. The dataset contains features pertaining to the ride like the time of day at which it occurred, the starting location of the ride, the destination, the distance, and the price. Other information like which rideshare provider was used, Uber or Lyft in this case, as well as the specific service used, UberXL, Shared, Lux, etc. was also provided. Our dataset also contained information about whether or not a surge multiplier had been applied. The dataset was also rich with weather data which included information like the temperature, humidity, pressure, visibility, cloud cover, UV Index, and a lot more. The dimensions of the dataset were 693071 rows by 57 columns. This meant that there were 693071 samples and 57 variables. It is important to note that some sample contained invalid values and some features were also removed during the cleaning process. This is why after the cleaning process, there were 637976 rows and 24 columns which meant that there were 637976 samples and 24 variables. 

### 2.2. Data Analysis
For our project, our group had two hypotheses tests that we wanted to conduct. The first regarded the price between Uber Pool and Lyft Shared. We wanted to know if the brand, Uber or Lyft, had an impact on the price of a ride. For this test, our Independent Variables were the brand of the ride, in this case Uber or Lyft. The reason why the type, Shared vs. Pool, was not an independent variable was due to the fact that these services were the same since they both provided shared rides to multiple people. Thus, this served as a control variable. The dependent variable in this case was the price since we wanted to see how the brand changed this quantity.

Our second hypothesis test that we wanted to test was if the difference in price was based on whether or not a ride took place during rush hour time or non rush hour time. In this case, the Rush Hour variable, which determined if a particular ride was occuring during a Rush Hour time or a Non-Rush Hour time was the independent variable while price was once again the dependent variable.

In our predictive model, our outcome variable is defined as the price. We intend to predict the price from the following feature variables: hour, day, month, source, destination, cab_type (Uber or Lyft), name (type of ride), distance, surge multiplier, temperature, apparentTemperature, short summary (of the weather), precipitation intensity, precipitation probability, humidity, wind speed, wind gust, visibility, dew point, pressure, cloud cover, UV index and ozone. We thought that distance and surge multiplier would definitely be important since a longer ride would mean that a person would spend more time traveling in a car which would cause the price to be higher. In addition, the surge multiplier was also important because we suspected that Uber and Lyft increased the prices of the rides at certain hours. Because of this, we thought that at particular times of day, on particular days in the month, these variables would have an impact on the target variable since we expected that the prices would be higher at times when people needed to use Ubers on weekends to spend time with others or on weekdays to get to work. Another key factor in the predictive model was definitely which type of ride was used (Pool, Lux, etc.). We anticipated that premium services like Uber Black, Uber Black SUV, Lux Black or Lux Black XL would cause the prices to be higher since the cars used for these services were more generally luxury vehicles. Lastly, we also thought it would be apt to take into account weather data since this could potentially influence the price of rides. On hot (high temperature) and humid days, people would probably be less likely to walk or use public transport and might prefer the comfort of an air-conditioned vehicle. We anticipated that the price would therefore be higher at these times. In addition, based on the intensity of rain or snow (precipitation intensity) or the likelihood of rain or snow occurring, people would be more likely to use the services of Uber or Lyft which would cause the price to be higher. For examples like these, we wanted to take the weather data into account since we reasoned that the weather could have an impact on the likelihood of a person using the services of Uber or Lyft. Uber and Lyft could use these metrics to modify the rates of the rides.  

The problem of predicting the price based on a multitude of factors was a supervised learning problem since we were using labeled data. The data already had a structure which included the hour, month, distance, surge multiplier, etc. and no patterns within the data needed to be uncovered. Instead, we were trying to predict the target value, in this case price, using a set of features. Within supervised learning, this project could be considered a Regression task because the price was the target variable and was numeric and continuous. 

Our group decided to choose three regression models: Multiple Linear Regression, Ridge Regression, and Lasso Regression. Multiple linear regression was a simple model that used ordinary least squares to find an equation for a line that minimized the distances between the actual data points and the predicted data points. The reason why multiple linear regression was used was due to the fact that our dataset had multiple features and we believed that these all needed to be taken into account in order to create a regression model that would be able to predict the cost of a ride in an Uber or Lyft. 

The second regression model used by our group was Ridge regression. Ridge regression was an algorithm that used L2 regularization to restrict coefficients to be as close to 0 as possible by minimizing the sum of the squared coefficients. The more restricted the coefficients, the smaller the impact of the coefficients on the predictions of the price. This in turn would combat the problem of overfitting. Yet another reason why Ridge Regression was used was to tune hyperparameters. Unfortunately, multiple linear regression did not allow for hyperparameters to be tuned, so this allowed us to have some more control over our model. Thus, for hyperparameter tuning as well as L2 regularization which combatted overfitting, the Ridge model was used.     

For similar reasons, our group also used Lasso regression. Lasso regression was an algorithm that also restricted the coefficients to be small in size, however this algorithm used L1 regularization. L1 regularization minimized the sum of the absolute values of the coefficients which meant that some of the coefficients could be 0. Thus, Lasso regression performed a form of feature selection on features that did not have a big impact on the target variable. L1 regularization also sought to combat the problem of overfitting like L2 regularization by restricting the size of the coefficients which was another advantage. Thus, Lasso regression was used because it performed a form of feature selection and addressed the problem of overfitting in addition to allowing us to tune hyperparameters. 

In [ ]:
# importing the rideshare dataset
import pandas as pd
rideshare_df = pd.read_csv("https://www.dropbox.com/s/1nyqv980whwhn2q/rideshare_kaggle.csv?dl=1")
modified_ride_share_df = rideshare_df

In [ ]:
modified_ride_share_df

id     timestamp  hour  day  month  \
0       424553bb-7174-41ea-aeb4-fe06d4f4b9d7  1.544953e+09     9   16     12   
1       4bd23055-6827-41c6-b23b-3c491f24e74d  1.543284e+09     2   27     11   
2       981a3613-77af-4620-a42a-0c0866077d1e  1.543367e+09     1   28     11   
3       c2d88af2-d278-4bfd-a8d0-29ca77cc5512  1.543554e+09     4   30     11   
4       e0126e1f-8ca9-4f2e-82b3-50505a09db9a  1.543463e+09     3   29     11   
...                                      ...           ...   ...  ...    ...   
693066  616d3611-1820-450a-9845-a9ff304a4842  1.543708e+09    23    1     12   
693067  633a3fc3-1f86-4b9e-9d48-2b7132112341  1.543708e+09    23    1     12   
693068  64d451d0-639f-47a4-9b7c-6fd92fbd264f  1.543708e+09    23    1     12   
693069  727e5f07-a96b-4ad1-a2c7-9abc3ad55b4e  1.543708e+09    23    1     12   
693070  e7fdc087-fe86-40a5-a3c3-3b2a8badcbda  1.543708e+09    23    1     12   

                   datetime          timezone            source  \
0       2018-12-16 09:30:07  America/New_York  Haymarket Square   
1       2018-11-27 02:00:23  America/New_York  Haymarket Square   
2       2018-11-28 01:00:22  America/New_York  Haymarket Square   
3       2018-11-30 04:53:02  America/New_York  Haymarket Square   
4       2018-11-29 03:49:20  America/New_York  Haymarket Square   
...                     ...               ...               ...   
693066  2018-12-01 23:53:05  America/New_York          West End   
693067  2018-12-01 23:53:05  America/New_York          West End   
693068  2018-12-01 23:53:05  America/New_York          West End   
693069  2018-12-01 23:53:05  America/New_York          West End   
693070  2018-12-01 23:53:05  America/New_York          West End   

          destination cab_type  ... precipIntensityMax uvIndexTime  \
0       North Station     Lyft  ...             0.1276  1544979600   
1       North Station     Lyft  ...             0.1300  1543251600   
2       North Station     Lyft  ...             0.1064  1543338000   
3       North Station     Lyft  ...             0.0000  1543507200   
4       North Station     Lyft  ...             0.0001  1543420800   
...               ...      ...  ...                ...         ...   
693066      North End     Uber  ...             0.0000  1543683600   
693067      North End     Uber  ...             0.0000  1543683600   
693068      North End     Uber  ...             0.0000  1543683600   
693069      North End     Uber  ...             0.0000  1543683600   
693070      North End     Uber  ...             0.0000  1543683600   

        temperatureMin  temperatureMinTime  temperatureMax  \
0                39.89          1545012000           43.68   
1                40.49          1543233600           47.30   
2                35.36          1543377600           47.55   
3                34.67          1543550400           45.03   
4                33.10          1543402800           42.18   
...                ...                 ...             ...   
693066           31.42          1543658400           44.76   
693067           31.42          1543658400           44.76   
693068           31.42          1543658400           44.76   
693069           31.42          1543658400           44.76   
693070           31.42          1543658400           44.76   

        temperatureMaxTime  apparentTemperatureMin  \
0               1544968800                   33.73   
1               1543251600                   36.20   
2               1543320000                   31.04   
3               1543510800                   30.30   
4               1543420800                   29.11   
...                    ...                     ...   
693066          1543690800                   27.77   
693067          1543690800                   27.77   
693068          1543690800                   27.77   
693069          1543690800                   27.77   
693070          1543690800                   27.77   

        apparentTemperatureMinTime  appar

<a id="3"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 3. RESULTS

#### Data Wrangling: Cleaning the dataframe
Remove columns that are redundant and invalid samples

In [ ]:
# cleaning and removing some columns that are redundant or bad data
columns_to_drop = ["id", "timestamp", "timezone", "datetime", "product_id", 
                   "latitude", "longitude", "long_summary", 
                  "windGustTime", "temperatureHigh", "temperatureHighTime", "temperatureLow",
                  "temperatureLowTime", "apparentTemperatureHigh", "apparentTemperatureHighTime", 
                  "apparentTemperatureLow", "apparentTemperatureLowTime", "icon", "windBearing", 
                  "visibility.1", "sunriseTime", "sunsetTime", "moonPhase", "precipIntensityMax",
                  "uvIndexTime", "temperatureMin", "temperatureMinTime", "temperatureMax", "temperatureMaxTime", 
                   "apparentTemperatureMin", "apparentTemperatureMinTime", "apparentTemperatureMax", 
                  "apparentTemperatureMaxTime"]

modified_ride_share_df = modified_ride_share_df.drop(columns_to_drop, axis = 1)

# removing invalid samples
modified_ride_share_df.dropna(inplace=True)

# fixes the indices after dropping
modified_ride_share_df.reset_index(drop=True, inplace=True)

In [ ]:
modified_ride_share_df

hour  day  month            source    destination cab_type  \
0          9   16     12  Haymarket Square  North Station     Lyft   
1          2   27     11  Haymarket Square  North Station     Lyft   
2          1   28     11  Haymarket Square  North Station     Lyft   
3          4   30     11  Haymarket Square  North Station     Lyft   
4          3   29     11  Haymarket Square  North Station     Lyft   
...      ...  ...    ...               ...            ...      ...   
637971    23    1     12          West End      North End     Uber   
637972    23    1     12          West End      North End     Uber   
637973    23    1     12          West End      North End     Uber   
637974    23    1     12          West End      North End     Uber   
637975    23    1     12          West End      North End     Uber   

                name  price  distance  surge_multiplier  ...  \
0             Shared    5.0      0.44               1.0  ...   
1                Lux   11.0      0.44               1.0  ...   
2               Lyft    7.0      0.44               1.0  ...   
3       Lux Black XL   26.0      0.44               1.0  ...   
4            Lyft XL    9.0      0.44               1.0  ...   
...              ...    ...       ...               ...  ...   
637971           WAV    9.5      1.00               1.0  ...   
637972        UberXL   13.0      1.00               1.0  ...   
637973         UberX    9.5      1.00               1.0  ...   
637974     Black SUV   27.0      1.00               1.0  ...   
637975      UberPool   10.0      1.00               1.0  ...   

        precipProbability  humidity windSpeed  windGust  visibility  dewPoint  \
0                     0.0      0.68      8.66      9.17      10.000     32.70   
1                     1.0      0.94     11.98     11.98       4.786     41.83   
2                     0.0      0.75      7.33      7.33      10.000     31.10   
3                     0.0      0.73      5.28      5.28      10.000     26.64   
4                     0.0      0.70      9.14      9.14      10.000     28.61   
...                   ...       ...       ...       ...         ...       ...   
637971                0.0      0.74      2.34      2.87       9.785     29.65   
637972                0.0      0.74      2.34      2.87       9.785     29.65   
637973                0.0      0.74      2.34      2.87       9.785     29.65   
637974                0.0      0.74      2.34      2.87       9.785     29.65   
637975                0.0      0.74      2.34      2.87       9.785     29.65   

        pressure  cloudCover  uvIndex  ozone  
0        1021.98        0.72        0  303.8  
1        1003.97        1.00        0  291.1  
2         992.28        0.03        0  315.7  
3        1013.73        0.00        0  291.1  
4         998.36        0.44        0  347.7  
...          ...         ...      ...    ...  
637971   1023.57        0.31        0  271.5  
637972   1023.57        0.31        0  271.5  
637973   1023.57        0.31        0  271.5  
637974   1023.57        0.31        0  271.5  
637975   1023.57        0.31        0  271.5  

[637976 rows x 24 columns]

### 3.2. Data Exploration

In [ ]:
import plotly.express as px

# groups the dataframe by the hour column and displays the means of each feature
hour_df = modified_ride_share_df.groupby(["hour"]).mean()

# creates a scatter plot of average price based on the hour and highlights the upper 
# 15% of data points
fig = px.scatter(hour_df, x=hour_df.index, y="price", title="Average Price By Hour", color="price", 
             color_continuous_scale = [(0, "blue"), (0.85, "blue"), (0.85, "red"), (1, "red")])

fig.update_yaxes(title="Average Price")

fig.show()

<center><img src = https://raw.githubusercontent.com/aneesh1/ds3000project/main/apbh2.png width=900></center>

### Figure 1: Average Price by Hour

The purpose of this graph was to display the average price at each hour in the day and to highlight the
upper 15% of data points. The reason why the upper 15% of data points were highlighted was to show that
these times represent "rush hour" times during which people are trying to get home from work or are going
out to for the night. It also is meant to highlight that at these times the average prices are significantly
higher than those of other times during the day. For example, at hours 8, 17, 20, 21, 22 the prices were deemed to be in this top 15% which is why they were later labeled as "Rush Hour" times.  

In [ ]:
# creates a dataframe that is grouped by the "name" or service provided by each of the rideshare providers
# services are defined as Black, Black SUV, UberPool, UberXL, Lux, Lux Black, etc.
name_df = modified_ride_share_df.groupby(["name"]).mean()

# created to map the different ride "names" to the company which provides these services
name_df["type"] = ["Uber", "Uber", "Lyft", "Lyft", "Lyft", "Lyft", "Lyft", "Lyft", "Uber", "Uber", "Uber", "Uber"]

# a bar plot that displays the different services (types of rides) provided by Uber and Lyft
# and displays their average prices. The graph is color-coded by the company providing the services
fig = px.bar(name_df, x=name_df.index, y="price", title="Average Price Based on Type of Ride", color="type", 
             color_discrete_sequence = ["black", "rgb(255, 0, 191)"], labels= {"name" : "Type of Ride"})

fig.update_yaxes(title="Average Price")

fig.show()

<center><img src = https://raw.githubusercontent.com/aneesh1/ds3000project/main/apbotor2.png width=1000></center>

### Figure 2: Average Price by Type of Ride

This bar plot was intended to show how the average price varies not only across Uber and Lyft, but also
across the type of rides within each rideshare provider. It can be seen from the graph that the most expensive
rides in both providers is the premium SUV option. In Uber, this is the "Black SUV" option and in Lyft this is the
"Lux Black XL" option. This graphic allows viewers to understand the price breakdowns within each provider. This graph showed that similar services provided by the rideshare services were relatively the same. For example, UberXL and LyftXL had roughly the same average price. UberX and Lyft also had average prices that were very similar.  

Constructing a dictionary to be converted into a heatmap that would show the number of rides from a particular location at a certain time.  

In [ ]:
# looks only at the hours and the source locations in order 
# to create a heatmap with the number of rides at a particular hour
# starting from a particular source
heatmap_df = modified_ride_share_df[["hour", "source"]]

# dictionary in which the keys represent the hours, and each index in the list
# represents how many rides are happening at a particular stations
values = {}

# maps the list indices for the above dictionary to specific locations
locations = {}
for i in range(0, 24):
    values[i] = [0] * 12

# creates the mappings for particular list indices to source locations
counter = 0
for i in heatmap_df["source"].value_counts().index.to_list():
    locations[i] = counter
    counter += 1

# goes through all of the rides and keeps track of which
# hour the ride is in and updates the values dictionary accordingly
counter = 0
for row in heatmap_df.itertuples():
    hour = row[1]
    source = row[2]
    values[hour][locations[source]] += 1

# converts values into a dataframe
to_show_df = pd.DataFrame.from_dict(values)

# changes the indices of the dataframe to be the names of the source locations
to_show_df.index = heatmap_df["source"].value_counts().index.to_list()

# creates a figure for the heatmap that shows the source locations and the number 
# of rides starting at these locations at particular hours
fig = px.imshow(to_show_df, title="Heatmap for Number of Trips by Location by Hour", 
                labels=dict(x="Hour", y="Source Location", color="Number of Trips"))

# updates the x-axes to read "Hour"
fig.update_xaxes(title="Hour")

# updates the x-axes to read "Source Location"
fig.update_yaxes(title="Source Location")
fig.show()

<center><img src = https://raw.githubusercontent.com/aneesh1/ds3000project/main/heatmap.png width=1000></center>

### Figure 3: Heatmap for Number of Trips by Location

Unlike the previous two graphics, this one provided a 3-D representation of data. The three components represented by this graph were: the hour, the source location, and the number of rides originating from a particular source location at a particular hour. The main purpose of this graph was to see which locations were people leaving from at particular times in order to get an idea of which times would be highly trafficked. This graph shows how people travel within the city from a particular location at a particular time. For example, it can be seen that at hour 17, or 5 P.M. the number of trips from the Financial District was among the highest in that specific hour. It can also be observed that at hour 0, a lot of trips were originating from Boston University which is most likely a lot of students are up late being involved in a lot of extra-curricular activities.  

#### Model Construction: Hypothesis Test 1

The purpose of this hypothesis test was to compare the price of an Uber Pool ride to a Lyft shared ride in order to determine if the difference between the two rides was statsitically significant. 

In [ ]:
from scipy import stats

# gets the price of Uber Pool rides
uber_prices =  modified_ride_share_df[modified_ride_share_df["name"] == "UberPool"]["price"]

# gets the price of Lyft Shared rides 
lyft_prices =  modified_ride_share_df[modified_ride_share_df["name"] == "Shared"]["price"]

# determines if a MannWhitney test ought to be run
runWhitney = False

print("-----------------------")
print("Independent Samples T-test")
print("-----------------------\n")

print("Assumption Checks")
print("-------------------\n")

# performs an assumption check using Levene's Test of Equality of Variances 
levene_result_1 = stats.levene(uber_prices, lyft_prices)

print("Assumption of Equality of Variances")
print(f'\t{levene_result_1}')

# determines if the assumption is true
if levene_results_1[1] > 0.05:
    print("\tAssumption is met. p > .05\n")
else:
    print("\tAssumption is not met. p < .05")
    print("\tIndependent Samples T-test cannot be conducted, so Mann-Whitney will be used instead.\n")
    # if the program gets here, the assumption check has failed so a non-parametric version of the
    # independent samples t-test, namely, the Mann-Whitney test needs to be performed
    runWhitney = True

print("Assumption of Normality")

# performs the Shapiro-Wilk test for normality on both of our distributions
shapiro_results_uber = stats.shapiro(uber_prices)
shapiro_results_lyft = stats.shapiro(lyft_prices)

print(f'\tUber Prices Shapiro-Wilk\'s Test Results: {shapiro_results_uber}')

if shapiro_results_uber[1] > 0.05:
    print('\t\tAssumption is met. p > .05\n')
else:
    print('\t\tAssumption is not met. p < .05')
    print("\t\tIndependent Samples T-test cannot be conducted, so Mann-Whitney will be used instead.\n")
    runWhitney=True

print(f'\tLyft Prices Shapiro-Wilk\'s Test Results: {shapiro_results_lyft}')

if shapiro_results_lyft[1] > 0.05:
    print('\t\tAssumption is met. p > .05\n')
else:
    print('\t\tAssumption is not met. p < .05')
    print("\t\tIndependent Samples T-test cannot be conducted, so Mann-Whitney will be used instead.\n")
    runWhitney=True

print(f'UberPool: n={uber_prices.count()}, M={uber_prices.mean()}, SE={uber_prices.sem()}')
print(f'Shared (Lyft): n={lyft_prices.count()}, M={lyft_prices.mean()}, SE={lyft_prices.sem()}\n')
# runs only if one of the assumption checks fails
if runWhitney:
    
    print("-----------------------")
    print("Mann-Whitney U-Test")
    print("-----------------------\n\n")
    
    print(f'Null Hypothesis: The probability of the price of one random sample from the Uber Pool')
    print(f'data being greater than that of the Shared (Lyft) data is equal to 50%\n')
    
    print(f'Alternate Hypothesis: The probability of the price of one random sample from the Uber Pool')
    print(f'data being greater than that of the Shared (Lyft) data is not equal to 50%\n')
    
    # runs a Mann-Whitney test
    mann_whitney_results = stats.mannwhitneyu(uber_prices, lyft_prices)
    
    # prints the results
    print(f'U={mann_whitney_results[0]}, p={mann_whitney_results[1]}\n')
    
    # interprets the results based on the pvalue and determines what to do 
    # with the null hypothesis
    if mann_whitney_results[1] > 0.05:
        print("Fail to reject the null hypothesis because p > alpha=0.05")
    else:
        print("Reject the null hypothesis because p < alpha=0.05")
else:
    ttestresults = stats.ttest_ind(lyft_prices, uber_prices)

    print(f'Null Hypothesis: There is no difference in the mean price of an Uber Pool ride in comparison to the Shared (Lyft) ride')
    print(f'Alternative Hypothesis: There is a difference in the mean price of an Uber Pool ride in comparison to the Shared (Lyft) ride')
    
    print(f'T={ttestresults[0]}, p={ttestresults[1]}, df = {uber_prices.count() + lyft_prices.count() - 2}\n')

    if ttestresults[1] > 0.05:
        print("Fail to reject the null hypothesis because p > alpha=0.05")
    else:
        print("Reject the null hypothesis because p < alpha=0.05")


-----------------------
Independent Samples T-test
-----------------------

Assumption Checks
-------------------

Assumption of Equality of Variances
	LeveneResult(statistic=387.04646997320197, pvalue=5.179236745067926e-86)
	Assumption is not met. p < .05
	Independent Samples T-test cannot be conducted, so Mann-Whitney will be used instead.

Assumption of Normality
	Uber Prices Shapiro-Wilk's Test Results: (0.9095629453659058, 0.0)
		Assumption is not met. p < .05
		Independent Samples T-test cannot be conducted, so Mann-Whitney will be used instead.

	Lyft Prices Shapiro-Wilk's Test Results: (0.9007724523544312, 0.0)
		Assumption is not met. p < .05
		Independent Samples T-test cannot be conducted, so Mann-Whitney will be used instead.

UberPool: n=55091, M=8.752500408415168, SE=0.008973662859992694
Shared (Lyft): n=51233, M=6.029892842503855, SE=0.009311937955354309

-----------------------
Mann-Whitney U-Test
-----------------------


Null Hypothesis: The probability of the price o

### Interpreting the Results from the T-test/Mann-Whitney Test for Hypothesis 1

We attempted to conduct an independent samples t test to compare the mean price of an UberPool ride and a Lyft Shared ride. However, because the assumptions of equality of variances and the assumptions of normality failed, we instead conducted a Mann Whitney U test. Results showed a statistically significant difference in the prices, with a U statistic of 459,827,728.5 and a p value of 0 which was less than the alpha value of 0.05. This means that we could reject the null hypothesis and conclude that the odds of picking an UberPool ride that had a greater price than a Lyft Shared ride was not equal to 50%. An examination of the prices revealed that UberPool (n = 55091, M = 8.75, SE = 0.00897) had a higher mean price than the Lyft shared (n = 51233, M = 6.03, SE = 0.0093). Overall, the statistically significant results show that we have confidence that an UberPool ride is more expensive than a Lyft shared ride in Boston.

### Data Wrangling: Feature Extraction
Performing some feature engineering to convert our "hour" data into a categorial variable that describes whether a particular ride is occuring at rush hour or not. This is a form of feature extraction since the Rush or Not Rush parameter is determined from the hour variable. This was done as a result of the Fig. 1 in the exploratory data analysis section

In [ ]:
# classifying the hours as rush or not rush (creating new features from existing ones)
rushOrNotRush = modified_ride_share_df["hour"].apply(lambda x: "Rush" if x in [8, 17, 20, 21, 22] else "Not Rush")

# drops the hour column since its values have been encoded
modified_ride_share_df.drop("hour", axis = 1, inplace=True)

# adds the newly created features into a new column called "Rush Hour"
modified_ride_share_df["Rush Hour"] = rushOrNotRush

In [ ]:
modified_ride_share_df

day  month            source    destination cab_type          name  \
0        16     12  Haymarket Square  North Station     Lyft        Shared   
1        27     11  Haymarket Square  North Station     Lyft           Lux   
2        28     11  Haymarket Square  North Station     Lyft          Lyft   
3        30     11  Haymarket Square  North Station     Lyft  Lux Black XL   
4        29     11  Haymarket Square  North Station     Lyft       Lyft XL   
...     ...    ...               ...            ...      ...           ...   
637971    1     12          West End      North End     Uber           WAV   
637972    1     12          West End      North End     Uber        UberXL   
637973    1     12          West End      North End     Uber         UberX   
637974    1     12          West End      North End     Uber     Black SUV   
637975    1     12          West End      North End     Uber      UberPool   

        price  distance  surge_multiplier  temperature  ...  humidity  \
0         5.0      0.44               1.0        42.34  ...      0.68   
1        11.0      0.44               1.0        43.58  ...      0.94   
2         7.0      0.44               1.0        38.33  ...      0.75   
3        26.0      0.44               1.0        34.38  ...      0.73   
4         9.0      0.44               1.0        37.44  ...      0.70   
...       ...       ...               ...          ...  ...       ...   
637971    9.5      1.00               1.0        37.05  ...      0.74   
637972   13.0      1.00               1.0        37.05  ...      0.74   
637973    9.5      1.00               1.0        37.05  ...      0.74   
637974   27.0      1.00               1.0        37.05  ...      0.74   
637975   10.0      1.00               1.0        37.05  ...      0.74   

       windSpeed  windGust  visibility  dewPoint  pressure  cloudCover  \
0           8.66      9.17      10.000     32.70   1021.98        0.72   
1          11.98     11.98       4.786     41.83   1003.97        1.00   
2           7.33      7.33      10.000     31.10    992.28        0.03   
3           5.28      5.28      10.000     26.64   1013.73        0.00   
4           9.14      9.14      10.000     28.61    998.36        0.44   
...          ...       ...         ...       ...       ...         ...   
637971      2.34      2.87       9.785     29.65   1023.57        0.31   
637972      2.34      2.87       9.785     29.65   1023.57        0.31   
637973      2.34      2.87       9.785     29.65   1023.57        0.31   
637974      2.34      2.87       9.785     29.65   1023.57        0.31   
637975      2.34      2.87       9.785     29.65   1023.57        0.31   

        uvIndex  ozone  Rush Hour  
0             0  303.8   Not Rush  
1             0  291.1   Not Rush  
2             0  315.7   Not Rush  
3             0  291.1   Not Rush  
4             0  347.7   Not Rush  
...         ...    ...        ...  
637971        0  271.5   Not Rush  
637972        0  271.5   Not Rush  
637973        0  271.5   Not Rush  
637974        0  271.5   Not Rush  
637975        0  271.5   Not Rush  

[637976 rows x 24 columns]

#### Model Construction: Hypothesis Test 2
The purpose of this hypothesis test was to compare the price of a ride during a Rush Hour time to a ride during a Non-Rush hour to determine if the difference between the rides at the two times was statsitically significant.

In [ ]:
# gets the price of rides that occur during Rush Hour
rush_hour_prices =  modified_ride_share_df[modified_ride_share_df["Rush Hour"] == "Rush"]["price"]

# gets the price of ridges that occur during Non-Rush hour times
not_rush_prices =  modified_ride_share_df[modified_ride_share_df["Rush Hour"] == "Not Rush"]["price"]

# determines if a MannWhitney test ought to be run
runWhitney2 = False

print("-----------------------")
print("Independent Samples T-test")
print("-----------------------\n")

print("Assumption Checks")
print("-------------------\n")

# performs an assumption check using Levene's Test of Equality of Variances 
levene_result_2 = stats.levene(rush_hour_prices, not_rush_prices)

print("Assumption of Equality of Variances")
print(f'\t{levene_result_2}')

# determines if the assumption is true
if levene_results_2[1] > 0.05:
    print("\tAssumption is met. p > .05\n")
else:
    print("\tAssumption is not met. p < .05")
    print("\tIndependent Samples T-test cannot be conducted, so Mann-Whitney will be used instead.\n")
    # if the program gets here, the assumption check has failed so a non-parametric version of the
    # independent samples t-test, namely, the Mann-Whitney test needs to be performed
    runWhitney2 = True

print("Assumption of Normality")

# performs the Shapiro-Wilk test for normality on both of our distributions
shapiro_results_rush = stats.shapiro(rush_hour_prices)
shapiro_results_not_rush = stats.shapiro(not_rush_prices)

print(f'\tRush Hour Prices Shapiro-Wilk\'s Test Results: {shapiro_results_rush}')

if shapiro_results_rush[1] > 0.05:
    print('\t\tAssumption is met. p > .05\n')
else:
    print('\t\tAssumption is not met. p < .05')
    print("\t\tIndependent Samples T-test cannot be conducted, so Mann-Whitney will be used instead.\n")
    runWhitney2=True

print(f'\tNot Rush Hour Prices Shapiro-Wilk\'s Test Results: {shapiro_results_not_rush}')

if shapiro_results_not_rush[1] > 0.05:
    print('\t\tAssumption is met. p > .05\n')
else:
    print('\t\tAssumption is not met. p < .05')
    print("\t\tIndependent Samples T-test cannot be conducted, so Mann-Whitney will be used instead.\n")
    runWhitney2=True

print(f'Rush Hour: n={rush_hour_prices.count()}, M={rush_hour_prices.mean()}, SE={rush_hour_prices.sem()}')
print(f'Not Rush Hour: n={not_rush_prices.count()}, M={not_rush_prices.mean()}, SE={not_rush_prices.sem()}\n')

# runs only if one of the assumption checks fails
if runWhitney2:
    
    print("-----------------------")
    print("Mann-Whitney U-Test")
    print("-----------------------\n\n")
    
    print(f'Null Hypothesis: The probability of the price of one random sample from the Rush Hour')
    print(f'data being greater than that of the Not Rush Hour data is equal to 50%\n')
    
    print(f'Alternate Hypothesis: The probability of the price of one random sample from the Rush Hour')
    print(f'data being greater than that of the Not Rush Hour data is not equal to 50%\n')
    
    # runs a Mann-Whitney test
    mann_whitney_results2 = stats.mannwhitneyu(rush_hour_prices, not_rush_prices)
    
    # prints the results
    print(f'U={mann_whitney_results2[0]}, p={mann_whitney_results2[1]}\n')
    
    # interprets the results based on the pvalue and determines what to do 
    # with the null hypothesis
    if mann_whitney_results2[1] > 0.05:
        print("Fail to reject the null hypothesis because p > alpha=0.05")
    else:
        print("Reject the null hypothesis because p < alpha=0.05")
else:
    ttestresults2 = stats.ttest_ind(rush_hour_prices, not_rush_prices)

    print(f'Null Hypothesis: There is no difference in the mean price of a Rush Hour ride in comparison to a Not Rush Hour ride.')
    print(f'Alternative Hypothesis: There is a difference in the mean price of a Rush Hour ride in comparison to a Not Rush Hour ride.')
    
    print(f'T={ttestresults2[0]}, p={ttestresults2[1]}, df = {rush_hour_prices.count() + not_rush_prices.count() - 2}\n')
    
    if ttestresults2[1] > 0.05:
        print("Fail to reject the null hypothesis because p > alpha=0.05")
    else:
        print("Reject the null hypothesis because p < alpha=0.05")

-----------------------
Independent Samples T-test
-----------------------

Assumption Checks
-------------------

Assumption of Equality of Variances
	LeveneResult(statistic=7.918996607320633, pvalue=0.004891989195525785)
	Assumption is not met. p < .05
	Independent Samples T-test cannot be conducted, so Mann-Whitney will be used instead.

Assumption of Normality
	Rush Hour Prices Shapiro-Wilk's Test Results: (0.9115809202194214, 0.0)
		Assumption is not met. p < .05
		Independent Samples T-test cannot be conducted, so Mann-Whitney will be used instead.

	Not Rush Hour Prices Shapiro-Wilk's Test Results: (0.9122379422187805, 0.0)
		Assumption is not met. p < .05
		Independent Samples T-test cannot be conducted, so Mann-Whitney will be used instead.

Rush Hour: n=127674, M=16.60206737471999, SE=0.026261518156938835
Not Rush Hour: n=510302, M=16.530879028496848, SE=0.013031943144442056

-----------------------
Mann-Whitney U-Test
-----------------------


Null Hypothesis: The probabilit

### Interpreting the Results from the T-test/Mann-Whitney Test for Hypothesis 2

We attempted to conduct an independent samples t test to compare the mean price of a Rush Hour ride and a non Rush Hour ride. However, because the assumptions of equality of variances and the assumptions of normality failed, we instead conducted a Mann Whitney U test. Results showed a statistically significant difference in the prices, with a U statistic of 32,475,888,288.5 and a p value of 0.044 which was less than the alpha value of 0.05. This means that we could reject the null hypothesis and conclude that the odds of picking a Rush Hour ride that had a greater price than a non Rush Hour ride was not equal to 50%. An examination of the prices revealed that Rush Hour rides (n = 127,674, M = 16.6, SE = 0.026) had a higher mean price than the non Rush Hour rides (n = 510,302, M = 16.53, SE = 0.013). Overall, the statistically significant results show that we have confidence that a Rush Hour ride is more expensive than a non Rush Hour ride.

#### Splitting the dataset into features and targets

In [ ]:
features, target = modified_ride_share_df.drop("price", axis = 1), modified_ride_share_df["price"]

In [ ]:
features

day  month            source    destination cab_type          name  \
0        16     12  Haymarket Square  North Station     Lyft        Shared   
1        27     11  Haymarket Square  North Station     Lyft           Lux   
2        28     11  Haymarket Square  North Station     Lyft          Lyft   
3        30     11  Haymarket Square  North Station     Lyft  Lux Black XL   
4        29     11  Haymarket Square  North Station     Lyft       Lyft XL   
...     ...    ...               ...            ...      ...           ...   
637971    1     12          West End      North End     Uber           WAV   
637972    1     12          West End      North End     Uber        UberXL   
637973    1     12          West End      North End     Uber         UberX   
637974    1     12          West End      North End     Uber     Black SUV   
637975    1     12          West End      North End     Uber      UberPool   

        distance  surge_multiplier  temperature  apparentTemperature  ...  \
0           0.44               1.0        42.34                37.12  ...   
1           0.44               1.0        43.58                37.35  ...   
2           0.44               1.0        38.33                32.93  ...   
3           0.44               1.0        34.38                29.63  ...   
4           0.44               1.0        37.44                30.88  ...   
...          ...               ...          ...                  ...  ...   
637971      1.00               1.0        37.05                37.05  ...   
637972      1.00               1.0        37.05                37.05  ...   
637973      1.00               1.0        37.05                37.05  ...   
637974      1.00               1.0        37.05                37.05  ...   
637975      1.00               1.0        37.05                37.05  ...   

       humidity  windSpeed  windGust  visibility  dewPoint  pressure  \
0          0.68       8.66      9.17      10.000     32.70   1021.98   
1          0.94      11.98     11.98       4.786     41.83   1003.97   
2          0.75       7.33      7.33      10.000     31.10    992.28   
3          0.73       5.28      5.28      10.000     26.64   1013.73   
4          0.70       9.14      9.14      10.000     28.61    998.36   
...         ...        ...       ...         ...       ...       ...   
637971     0.74       2.34      2.87       9.785     29.65   1023.57   
637972     0.74       2.34      2.87       9.785     29.65   1023.57   
637973     0.74       2.34      2.87       9.785     29.65   1023.57   
637974     0.74       2.34      2.87       9.785     29.65   1023.57   
637975     0.74       2.34      2.87       9.785     29.65   1023.57   

        cloudCover  uvIndex  ozone  Rush Hour  
0             0.72        0  303.8   Not Rush  
1             1.00        0  291.1   Not Rush  
2             0.03        0  315.7   Not Rush  
3             0.00        0  291.1   Not Rush  
4             0.44        0  347.7   Not Rush  
...            ...      ...    ...        ...  
637971        0.31        0  271.5   Not Rush  
637972        0.31        0  271.5   Not Rush  
637973        0.31        0  271.5   Not Rush  
637974        0.31        0  271.5   Not Rush  
637975        0.31        0  271.5   Not Rush  

[637976 rows x 23 columns]

In [ ]:
target

0          5.0
1         11.0
2          7.0
3         26.0
4          9.0
          ... 
637971     9.5
637972    13.0
637973     9.5
637974    27.0
637975    10.0
Name: price, Length: 637976, dtype: float64

#### Model Construction: Splitting the data into training and testing sets 

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer

# splits the data using the train-test split approach
X_train, X_test, y_train, y_test = train_test_split(features, target, random_state=3000)

In [ ]:
X_train

day  month                   source              destination cab_type  \
326498   16     12  Northeastern University         Theatre District     Uber   
587843   15     12            North Station  Northeastern University     Lyft   
567721   17     12  Northeastern University                 Back Bay     Lyft   
420338   29     11            South Station            North Station     Uber   
115291   26     11            South Station       Financial District     Uber   
...     ...    ...                      ...                      ...      ...   
211545   29     11                   Fenway            North Station     Lyft   
624468   14     12       Financial District        Boston University     Lyft   
508853   16     12                 Back Bay                North End     Uber   
378798    3     12            South Station         Theatre District     Uber   
220668   16     12         Haymarket Square                 West End     Lyft   

             name  distance  surge_multiplier  temperature  \
326498      Black      1.91               1.0        43.06   
587843  Lux Black      3.34               1.0        43.94   
567721        Lux      1.48               1.0        38.75   
420338      Black      2.29               1.0        41.39   
115291   UberPool      0.54               1.0        44.15   
...           ...       ...               ...          ...   
211545    Lyft XL      3.23               1.0        38.35   
624468    Lyft XL      5.34               1.0        45.82   
508853        WAV      3.17               1.0        42.85   
378798   UberPool      1.30               1.0        47.87   
220668     Shared      0.58               1.0        42.84   

        apparentTemperature  ... humidity  windSpeed  windGust  visibility  \
326498                38.26  ...     0.71       8.12     13.44      10.000   
587843                41.06  ...     0.95       5.05      8.47       6.960   
567721                33.51  ...     0.84       7.19      8.88       8.325   
420338                36.08  ...     0.58       8.41     15.95       9.994   
115291                42.54  ...     0.93       3.53      3.90       8.561   
...                     ...  ...      ...        ...       ...         ...   
211545                32.50  ...     0.64       8.14     11.55       9.960   
624468                43.92  ...     0.53       4.16      4.16       9.874   
508853                38.10  ...     0.73       7.92     12.70       9.984   
378798                47.87  ...     0.93       2.77      4.54       2.069   
220668                37.91  ...     0.72       8.28     10.42       9.858   

        dewPoint  pressure  cloudCover  uvIndex  ozone  Rush Hour  
326498     34.25   1015.00        1.00        0  322.7   Not Rush  
587843     42.68   1023.46        1.00        0  296.2   Not Rush  
567721     34.41   1000.46        1.00        1  335.8   Not Rush  
420338     27.59   1006.23        0.22        1  331.8   Not Rush  
115291     42.14   1014.19        1.00        1  295.0   Not Rush  
...          ...       ...         ...      ...    ...        ...  
211545     27.39   1004.90        0.29        0  341.8   Not Rush  
624468     29.78   1029.37        0.39        0  289.0   Not Rush  
508853     34.93   1014.49        0.96        0  322.3       Rush  
378798     46.01   1000.30        1.00        0  300.5   Not Rush  
220668     34.27   1014.20        0.92        0  322.1       Rush  

[478482 rows x 23 columns]

#### Data Wrangling: One-Hot Encoding, Min-Max Scaling
Here, we scale the continuous variables and one-hot encode the categorical variables

In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import OneHotEncoder

# provides a list of all of the features, continuous and categorical both included
all_features = features.columns.to_list()

# list of all of the categorical features
categorical_features = ["source", "destination", "cab_type", "name", "short_summary", "Rush Hour"]

# creates a list of the continuous features by filtering out the categorical features from the list 
# of all the features
continuous_features = list(filter(lambda x: x not in categorical_features, all_features))

# tuple contianing the MinMaxScaler which is to be scaled on the continuous features
min_max = ("min_max", MinMaxScaler(), continuous_features)

# tuple containing the OneHotEncoder which is to be used to encode the categorical variables
# so that the overall dataset becomes numeric
one_hot_encoder = ("one-hot", OneHotEncoder(sparse=False), categorical_features)

# scales ONLY the continuous features as specified by the tuple and 
# encodes ONLY the categorial features as specified by the tuple
min_max_scaler = ColumnTransformer(transformers = [min_max, one_hot_encoder])

# encodes ONLY the categorial features specified by the tuple
one_hot_encode = ColumnTransformer(transformers = [one_hot_encoder], remainder='passthrough')

# training data with one-hot encoding applied
X_train_encoded = one_hot_encode.fit_transform(X_train)
# testing data with one-hot encoding applied
X_test_encoded = one_hot_encode.fit_transform(X_test)

# training data with one-hot encoding and min-max scaling applied
X_train_min_max = min_max_scaler.fit_transform(X_train)
# testing data with one-hot encoding and min-max scaling applied
X_test_min_max = min_max_scaler.fit_transform(X_test)

In [ ]:
pd.DataFrame(X_train_min_max)

0    1         2    3         4         5        6     7   \
0       0.517241  1.0  0.241071  0.0  0.630384  0.579508  0.00000  0.00   
1       0.482759  1.0  0.423469  0.0  0.653354  0.641606  0.00000  0.00   
2       0.551724  1.0  0.186224  0.0  0.517880  0.474163  0.00000  0.00   
3       0.965517  0.0  0.289541  0.0  0.586792  0.531160  0.00000  0.00   
4       0.862069  0.0  0.066327  0.0  0.658836  0.674429  0.05114  0.86   
...          ...  ...       ...  ...       ...       ...      ...   ...   
478477  0.965517  0.0  0.409439  0.0  0.507439  0.451763  0.00000  0.00   
478478  0.448276  1.0  0.678571  0.0  0.702428  0.705034  0.00000  0.00   
478479  0.517241  1.0  0.401786  0.0  0.624902  0.575959  0.00000  0.00   
478480  0.068966  1.0  0.163265  0.0  0.755938  0.792637  0.00000  0.00   
478481  0.517241  1.0  0.071429  0.0  0.624641  0.571745  0.00000  0.00   

              8         9   ...   56   57   58   59   60   61   62   63   64  \
0       0.568966  0.527148  ...  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0   
1       0.982759  0.316151  ...  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0   
2       0.793103  0.463230  ...  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0   
3       0.344828  0.547079  ...  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   
4       0.948276  0.211684  ...  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
...          ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
478477  0.448276  0.528522  ...  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   
478478  0.258621  0.254983  ...  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   
478479  0.603448  0.513402  ...  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
478480  0.948276  0.159450  ...  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0   
478481  0.586207  0.538144  ...  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   

         65  
0       0.0  
1       0.0  
2       0.0  
3       0.0  
4       0.0  
...     ...  
478477  0.0  
478478  0.0  
478479  1.0  
478480  0.0  
478481  1.0  

[478482 rows x 66 columns]

#### Data Wrangling: Feature Selection

Model-based feature selection is performed to select the most important features in our dataset 

In [ ]:
from sklearn.feature_selection import SelectFromModel

# model-based feature selection that uses the DecisionTreeRegressor to select features
# whose weighted importance is greater than the median weighted importance
model_based = SelectFromModel(DecisionTreeRegressor(random_state = 3000), threshold='median')

# resets the indicies of the training data so merge operations can perform w/o data loss
X_train.reset_index()
# resets the indicies of the testing data so merge operations can perform w/o data loss
X_test.reset_index()

# scales and transforms the continuous training features using the min-max scaler
X_train_selected = min_max_scaler.fit_transform(X_train)

# performs feature selection on the scaled training data (X_train_selected)
# categorial variables unfortunately get dropped, but are later added back
partial_X_train_selected = model_based.fit_transform(pd.DataFrame(X_train_selected).iloc[:, 0:17], y_train)

# scales and transforms the continuous testing features using the min-max scaler
X_test_selected = min_max_scaler.fit_transform(X_test)

# performs feature selection on the scaled testing data (X_test_selected)
# categorial variables unfortunately get dropped, but are later added back
partial_X_test_selected = model_based.fit_transform(pd.DataFrame(X_test_selected).iloc[:, 0:17], y_test)

# combines the continuous training features that have been selected with the 
# one-hot encoded categorial training features
X_train_selected = pd.merge(pd.DataFrame(partial_X_train_selected), pd.DataFrame(X_train_selected).iloc[:, 17:], 
                           left_index=True, right_index=True, how="inner")

# combines the continuous testing features that have been selected with the 
# one-hot encoded categorial testing features
X_test_selected = pd.merge(pd.DataFrame(partial_X_test_selected), pd.DataFrame(X_test_selected).iloc[:, 17:], 
                           left_index=True, right_index=True, how="inner")


# creates a list of the continuous features that remain after model-based feature selection 
# has beenp erformed
filtered_continuous = []
for mask, var in zip(model_based.get_support(), continuous_features):
    if mask:
        filtered_continuous.append(var)

# one-hot encoder object
encoder = OneHotEncoder(sparse = False)

# dataframe containing the categorical variables that need to be encoded
features_to_encode = modified_ride_share_df[["source", "destination", "cab_type", "name", "short_summary", "Rush Hour"]]

# contains a dataframe with the encoded features
encoded_features = encoder.fit_transform(features_to_encode)

# list that contains the names of the features after being encoded
listOfColumns = encoder.get_feature_names()

# the features names are of the form xi_[value]
# to make that more readable, these mappings are used to relate xi to the actual names
mappings = {
    "x0" : "source",
    "x1" : "destination",
    "x2" : "cab_type",
    "x3" : "name",
    "x4" : "short_summary",
    "x5" : "Rush_Hour"
}

# maps the list of the feature names so that their actual names are shown 
# as opposed to being in the form xi_value
for i in range(0, len(listOfColumns)):
    new_str = mappings[listOfColumns[i][0:2]] + listOfColumns[i][2:] 
    new_str = new_str.replace(" ", "")
    listOfColumns[i] = new_str

# listing containing all of the column names, which consists of the selected continuous features
# and the encoded features
new_columns = filtered_continuous + listOfColumns.tolist()

# sets the column names of the training data to the selected continuous 
# features and the encoded feature names
X_train_selected.columns = new_columns

# sets the column names of the testing data to the selected continuous 
# features and the encoded feature names
X_test_selected.columns = new_columns

# prints the continuous features selected by the model-based feature selection
print(f'Continuous features selected by model-based feature selection: {filtered_continuous}')

Continuous features selected by model-based feature selection: ['distance', 'surge_multiplier', 'temperature', 'windSpeed', 'windGust', 'visibility', 'dewPoint', 'pressure', 'ozone']


In [ ]:
X_train_selected

distance  surge_multiplier  temperature  windSpeed  windGust  \
0       0.241071               0.0     0.579508   0.527148  0.477883   
1       0.423469               0.0     0.641606   0.316151  0.289981   
2       0.186224               0.0     0.474163   0.463230  0.305482   
3       0.289541               0.0     0.531160   0.547079  0.572779   
4       0.066327               0.0     0.674429   0.211684  0.117202   
...          ...               ...          ...        ...       ...   
478477  0.409439               0.0     0.451763   0.528522  0.406427   
478478  0.678571               0.0     0.705034   0.254983  0.127032   
478479  0.401786               0.0     0.575959   0.513402  0.449905   
478480  0.163265               0.0     0.792637   0.159450  0.141399   
478481  0.071429               0.0     0.571745   0.538144  0.363705   

        visibility  dewPoint  pressure     ozone  source_BackBay  ...  \
0         1.000000  0.645203  0.567004  0.486758             0.0  ...   
1         0.672520  0.827355  0.745259  0.244749             0.0  ...   
2         0.819563  0.648660  0.260641  0.606393             0.0  ...   
3         0.999354  0.501296  0.382217  0.569863             0.0  ...   
4         0.844985  0.815687  0.549937  0.233790             0.0  ...   
...            ...       ...       ...       ...             ...  ...   
478477    0.995691  0.496975  0.354193  0.661187             0.0  ...   
478478    0.986427  0.548617  0.869785  0.178995             0.0  ...   
478479    0.998276  0.659896  0.556258  0.483105             1.0  ...   
478480    0.145643  0.899309  0.257269  0.284018             0.0  ...   
478481    0.984703  0.645635  0.550147  0.481279             0.0  ...   

        short_summary_Drizzle  short_summary_Foggy  short_summary_LightRain  \
0                         0.0                  0.0                      0.0   
1                         0.0                  0.0                      0.0   
2                         0.0                  0.0                      0.0   
3                         0.0                  0.0                      0.0   
4                         1.0                  0.0                      0.0   
...                       ...                  ...                      ...   
478477                    0.0                  0.0                      0.0   
478478                    0.0                  0.0                      0.0   
478479                    0.0                  0.0                      0.0   
478480                    0.0                  0.0                      0.0   
478481                    0.0                  0.0                      0.0   

        short_summary_MostlyCloudy  short_summary_Overcast  \
0                              0.0                     1.0   
1                              0.0                     1.0   
2                              0.0                     1.0   
3                              0.0                     0.0   
4                              0.0                     0.0   
...                            ...                     ...   
478477                         0.0                     0.0   
478478                         0.0                     0.0   
478479                         0.0                     1.0   
478480                         0.0                     1.0   
478481                         0.0                     1.0   

        short_summary_PartlyCloudy  short_summary_PossibleDrizzle  \
0                              0.0                            0.0   
1                              0.0                            0.0   
2                              0.0                            0.0   
3                              1.0                            0.0   
4                              0.0                            0.0   
...                            ...                            ...   
478477                         1.0                            0.0   
478478              

#### Model Evaluation and Testing: Multiple Linear Regression Model Training and Testing
The multiple linear regression model is trained and tested. It is evaluated on using R^2 score and Mean-Squared Error

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error

print("Multiple Linear Regression Results\n")

# trains the multiple linear regression model
multiple_linear = LinearRegression().fit(X=X_train_min_max, y=y_train)

# prints the results before feature selection is performed
print("Results before feature selection:")

# performs an R2 on the the multiple linear regression model before feature selection is performed on the training data
print(f'R2 Score on the training data: {r2_score(y_train, multiple_linear.predict(X_train_min_max))}')

# performs an R2 on the the multiple linear regression model before feature selection is performed on the testing data
print(f'R2 Score on the testing data: {r2_score(y_test, multiple_linear.predict(X_test_min_max))}\n')

# determines the Mean Squared Error on the the multiple linear regression model before feature selection is performed 
# on the training data
print(f'Mean squared error on training data: {mean_squared_error(y_train, multiple_linear.predict(X_train_min_max))}')

# determines the Mean Squared Error on the the multiple linear regression model before feature selection is performed 
# on the testing data
print(f'Mean squared error on testing data: {mean_squared_error(y_test, multiple_linear.predict(X_test_min_max))}')

# fits the multiple linear regression after feature selection
multiple_linear = LinearRegression().fit(X=X_train_selected, y=y_train)

# prints the results after the selection
print("\nResults after feature selection:")

# performs an R2 on the training data with the model after feature selection
print(f'R2 Score on the training data: {r2_score(y_train, multiple_linear.predict(X_train_selected))}')

# performs an R2 on the testing data with the model after feature selection
print(f'R2 Score on the testing data: {r2_score(y_test, multiple_linear.predict(X_test_selected))}\n')

# determines the mean squared error on the model after feature selection on the training data
print(f'Mean squared error on training data: {mean_squared_error(y_train, multiple_linear.predict(X_train_selected))}')

# determines the mean squared error on the model after feature selection on the testing data
print(f'Mean squared error on testing data: {mean_squared_error(y_test, multiple_linear.predict(X_test_selected))}')

Multiple Linear Regression Results

Results before feature selection:
R2 Score on the training data: 0.9288018377052939
R2 Score on the testing data: 0.9280462373805496

Mean squared error on training data: 6.1876210059114936
Mean squared error on testing data: 6.263782854877911

Results after feature selection:
R2 Score on the training data: 0.9287995796182996
R2 Score on the testing data: 0.9280451141663797

Mean squared error on training data: 6.187817249551341
Mean squared error on testing data: 6.263880633915496


#### Model Evaluation, Optimization, and Testing: Ridge Regression Model Training and Testing
The multiple linear regression model is trained and tested. It is evaluated on using R^2 score and Mean-Squared Error. Grid Search with Cross Validation was performed in order to fine-tune models and determine optimal hyperparameters. This was done to restrict coefficients which would in turn avoid overfitting. An explanation of the meaning of R^2 and Mean Squared Error is provided at the end of this section. 

In [ ]:
from sklearn.linear_model import Ridge
from sklearn.model_selection import GridSearchCV

# different alpha parameters to try and tune the Ridge regression model 
param_grid_ridge = {'alpha':[0.001, 0.01, 1, 7, 10, 29]}

# performing a grid search with cross validation without feature selection to assess model performance
grid_search_ridge = GridSearchCV(Ridge(), param_grid_ridge, cv = 5)

# fits the Ridge regression model on the training data without feature selection
grid_search_ridge.fit(X=X_train_min_max, y=y_train)

print("Ridge Regression Results\n")
print("Results before feature selection")

# gets the best alpha value after grid search on the training data without feature selection 
# is complete and the model has been refit
best_alpha_value = grid_search_ridge.best_params_["alpha"]

# prints the best alpha value after grid search with cross validation has completed
print(f'Best Alpha Value: {best_alpha_value}')

# prints the best score obtained in the cross validation process
print(f'Best Score: {grid_search_ridge.best_score_}')

# prints the R2 value of the training data without feature selection
# using the Ridge model that has been refit
print(f'R2 Score on training with Cross-Validation: {grid_search_ridge.score(X_train_min_max, y_train)}')

# prints the R2 value of the testing data without feature selection
# using the Ridge model that been refit with the optimized alpha parameter
print(f'R2 Score on testing with Cross-Validation: {grid_search_ridge.score(X_test_min_max, y_test)}\n')

# prints the mean squared error (MSE) on the training data without feature selection
# using the Ridge model that has been refit with the optimized alpha parameter
print(f'Mean squared error on training data: {mean_squared_error(y_train, grid_search_ridge.predict(X_train_min_max))}')

# prints the mean squared error (MSE) on the testing data without feature selection 
# using the Ridge model that has been refit with the optimized alpha parameter
print(f'Mean squared error on testing data: {mean_squared_error(y_test, grid_search_ridge.predict(X_test_min_max))}\n\n')

# fits the Ridge regression model on the training data after, or with, feature selection applied
grid_search_ridge.fit(X=X_train_selected, y=y_train)
print("Results after feature selection")

# best alpha value after the cross validation process completes with feature selection applied
best_alpha_value = grid_search_ridge.best_params_["alpha"]
print(f'Best Alpha Value: {best_alpha_value}')

# prints the best score in the cross validation process with feature selection applied
print(f'Best Score: {grid_search_ridge.best_score_}')

# prints the R2 value on the training data after feature selection and after the Ridge model has been refit 
print(f'R2 Score on training with Cross-Validation: {grid_search_ridge.score(X_train_selected, y_train)}')

# prints the R2 value on the testing data after feature selection and after the Ridge model has been refit 
print(f'R2 Score on testing with Cross-Validation: {grid_search_ridge.score(X_test_selected, y_test)}\n')

# prints the mean squared error on the training data after feature selection and after the Ridge model has been refit 
print(f'Mean squared error on training data: {mean_squared_error(y_train, grid_search_ridge.predict(X_train_selected))}')

# prints the mean squared error on the testing data after feature selection and after the Ridge model has been refit 
print(f'Mean squared error on testing data: {mean_squared_error(y_test, grid_search_ridge.predict(X_test_selected))}\n\n')

Ridge Regression Results

Results before feature selection
Best Alpha Value: 1
Best Score: 0.9287716240346849
R2 Score on training with Cross-Validation: 0.9288019640705646
R2 Score on testing with Cross-Validation: 0.928047045263252

Mean squared error on training data: 6.1876100238808105
Mean squared error on testing data: 6.263712526355314


Results after feature selection
Best Alpha Value: 0.01
Best Score: 0.9287719107346135
R2 Score on training with Cross-Validation: 0.9287996264738311
R2 Score on testing with Cross-Validation: 0.9280452133594743

Mean squared error on training data: 6.187813177476156
Mean squared error on testing data: 6.263871998870106




#### Model Evaluation, Optimization, and Testing: Lasso Regression Model Training and Testing
The Lasso regression model is trained and tested. It is evaluated on using R^2 score and Mean-Squared Error. Grid Search with Cross Validation was performed in order to fine-tune models and determine optimal hyperparameters. This was done to restrict coefficients which would in turn avoid overfitting. An explanation of the meaning of R^2 and Mean Squared Error is provided at the end of this section. 

In [ ]:
from sklearn.linear_model import Lasso

# different alpha values to try for Lasso regression
param_grid_lasso = {'alpha':[0.001, 0.01, 0.1, 1, 10]}

# sets up the Lasso regression for Grid Search with Cross Validation
grid_search_lasso = GridSearchCV(Lasso(), param_grid_lasso, cv = 5)

# fits the Lasso model using the training data without feature selection applied 
grid_search_lasso.fit(X=X_train_min_max, y=y_train)

print("Lasso Regression Results\n")
print("Results before feature selection")

# prints the best alpha value after the Grid Search with Cross Validaton process has been completed
best_alpha_value = grid_search_lasso.best_params_["alpha"]
print(f'Best Alpha Value: {best_alpha_value}')

# prints the best score received through the cross validation process without feature selection applied 
print(f'Best Score: {grid_search_lasso.best_score_}')

# prints the R2 score on the training data without feature selection applied 
print(f'R2 Score on training with Cross-Validation: {grid_search_lasso.score(X_train_min_max, y_train)}')

# prints the R2 score on the testing data without feature selection applied 
print(f'R2 Score on testing with Cross-Validation: {grid_search_lasso.score(X_test_min_max, y_test)}\n')

# prints the mean squared error on the training data without feature selection applied 
print(f'Mean squared error on training data: {mean_squared_error(y_train, grid_search_lasso.predict(X_train_min_max))}')

# prints the mean squared error on the testing data without feature selection applied 
print(f'Mean squared error on testing data: {mean_squared_error(y_test, grid_search_lasso.predict(X_test_min_max))}\n\n')

# fits the Lasso regression model to the training data with feature selection applied
grid_search_lasso.fit(X=X_train_selected, y=y_train)
print("Results after feature selection")

# prints the best alpha value that generated the highest score
best_alpha_value = grid_search_lasso.best_params_["alpha"]
print(f'Best Alpha Value: {best_alpha_value}')

# prints the best score obtained in the grid search with cross validation process
print(f'Best Score: {grid_search_lasso.best_score_}')

# prints the R2 score on the training data with feature selection applied 
print(f'R2 Score on training with Cross-Validation: {grid_search_lasso.score(X_train_selected, y_train)}')

# prints the R2 score on the testing data with feature selection applied 
print(f'R2 Score on testing with Cross-Validation: {grid_search_lasso.score(X_test_selected, y_test)}\n')

# prints the mean squared error on the training data with feature selection applied 
print(f'Mean squared error on training data: {mean_squared_error(y_train, grid_search_lasso.predict(X_train_selected))}')

# prints the mean squared error on the testing data with feature selection applied 
print(f'Mean squared error on testing data: {mean_squared_error(y_test, grid_search_lasso.predict(X_test_selected))}\n\n')

Lasso Regression Results

Results before feature selection
Best Alpha Value: 0.001
Best Score: 0.9287635683851615
R2 Score on training with Cross-Validation: 0.9287852839067386
R2 Score on testing with Cross-Validation: 0.9280371612240208

Mean squared error on training data: 6.189059647420881
Mean squared error on testing data: 6.264572960517743


Results after feature selection
Best Alpha Value: 0.001
Best Score: 0.9287635723786118
R2 Score on training with Cross-Validation: 0.9287852839067386
R2 Score on testing with Cross-Validation: 0.9280371612240208

Mean squared error on training data: 6.189059647420881
Mean squared error on testing data: 6.264572960517743




### Explanation of R^2 and Mean-Squared Error
Because we were performing a regression task, we decided to use the R^2 metric and the Mean-Squared Error since these would provide insight into the model's performance. The R^2 value represented how much of the variance in the dataset our model could account for. It stands to reason that an R^2 or 0.928 meant that our model could explain about 92.8% of the variance. The mean-squared error metric provided insight into the error between our predictions and the actual data. The mean squared error was the average of the sum of the squares of the errors. Thus, if we were to take the square root of the mean squared error, we could interpret that to mean that our model was off by an average of $2.51.  

<a id="4"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 4. DISCUSSION

We compared three algorithms: Multiple Linear Regression, Ridge Regression, and Lasso Regression. We evaluated the performance of these three algorithms before and after conducting model-based feature selection, to see if only choosing the most important features would lead to a higher R-squared value and mean squared error. Within our Ridge Regression and Lasso Regression algorithms, we used gridsearchCV() in order to obtain the alpha parameter that would yield the best R-squared score and mean squared error on the testing and training data. 

In our Multiple Linear Regression model, the R-squared score and mean squared error on the training and testing data was barely affected by the use of model-based feature selection. The performance of the model after feature selection negligibly decreased, however this is not enough to warrant that model-based feature selection impacted the performance of our Multiple Linear Regression model.

For our Ridge Regression model, the best alpha value before model-based feature selection was 1, however after model-based feature selection, that alpha value changed to 0.01. However, the difference between the mean cross-validated R-squared score of the best alpha value and the R-squared values of the training set and testing set were nearly identical between the two alpha values. Furthermore, the mean squared error on the training and testing data had a slight increase after applying model-based feature selection. Thus, we can conclude that feature selection does not impact the performance of our Ridge regression model.

Regarding the Lasso Regression model, the best alpha value before and after model-based feature selection was the same: 0.001. As a result, the R-squared score and mean squared error on the training and testing sets were identical before and after the model-based feature selection. The mean cross-validated R-squared score of this alpha value differs extremely slightly though, because the 5 folds created from the data vary between each call to best_score_. Hence, we can conclude that model-based feature selection does not impact the performance of our Lasso regression model.

In a general sense, all three algorithms boasted the same performance. However, if we scrutinize past the sixth or seventh decimal points, we can see that Ridge Regression has the best performance: having the highest R-squared score and lowest MSE. In other words, the Ridge Regression model fits and predicts the data better than a standard Multiple Linear Regression and Lasso regression. There is no indication of overfitting or underfitting because the R-squared score is very high, and nearly identical, for both the training and testing datasets. Our R-squared score was approximately .93, which is very close to 1, thus indicating that our models can make nearly perfect predictions. Furthermore, the mean squared error was approximately 6.188 on the training data, and 6.264 on the testing data. This means that on average, our predicted price deviated about $2.50 from the actual price of a given rideshare.

After a careful analysis of these three algorithms, a Ridge Regression should be used for our predictive model, however there is no standout winner in terms of the best predictive model. It makes sense that Ridge Regression had the strongest performance because when dealing with datasets that have a lot of important features, it is best to use Ridge Regression because this model uses L2 regularization which ensures that all features are taken into account, even if they have a small influence on the outcome variable. This is very much unlike Lasso regression since this model has the ability to potentially set the coefficients of certain features to 0, meaning that they will have absolutely no impact on the outcome variable. 

Based on our findings, we can use the features in our dataset to predict the price of a rideshare using the algorithms we’ve applied. Although we initially thought weather would have a large factor in determining the price of a rideshare, the more common and intuitive features such as distance, surge multiplier, and the type of rideshare (for instance a Black Uber SUV, or Lyft XL) were more useful in predicting the target variable. Our model could still use some improvement, but it would prove to be fairly reliable if used in the real world.

In addition, the results from our hypothesis tests provide an indication that the differences in price based on which rideshare service was used was statistically significant because we rejected the null hypothesis. This means that we can be confident that there is a difference in price between the services provided by Uber and Lyft. Thus in this case, having statisitcally significant results meant led to the conclusion that there was a difference in price based on which service was used. 

In our second hypothesis test, we tried to discern if the relationship between use of a car during a time that was "rush hour" or a time that was "non-rush hour" and price was statistically significant. After performing a Mann-Whitney test, we rejected the null hypothesis which meant that there was indeed a statistically significant relationship between the price of a ride at a rush hour time and at a non-rush hour time. Hence, in this case being statistically significant meant that the time of day had an impact on the price of a ride. 

Overall, we think that our final project was successful, but future work could still be done on this topic. For instance, different regressors could be fit on our data. In fact, we discussed a random tree regressor, which yielded a better R-squared score and lower MSE, however it was outside of the scope of this class. In addition, we could perform more analyses on a larger dataset, for instance with more ridesharing apps, or more locations outside of Boston.

<a id="5"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## References
Muthhukumar, M. (2019, July 31). Presenting to you the Lyft and Uber Price Predictor known as the Ride- Hailer. [Web log post]. Retrieved December 11, 2020, from https://medium.com/@malaiiyy/presenting-to-you-the-lyft-and-uber-price-predictor-known-as-the-ride-hailer-47dea6dc7a9a

Uber, Data Science. (2018). Dynamic Pricing and Matching in Ride-Sharing [Press release]. Retrieved December 11, 2020, from https://www.naefrontiers.org/184199/Abstract

Yatnalkar, G. (2019). A Machine Learning Recommender Model for Ride Sharing Based on Rider Characteristics and User Threshold Time (Unpublished master's thesis). Marshall University. Retrieved December 11, 2020, from https://mds.marshall.edu/cgi/viewcontent.cgi?article=2263&context=etd

### CONTRIBUTIONS

* Siddhant Dosi - Section 1
* Aneesh Atri - Section 2
* Justin Lau - Section 4
* All Group Members (Trio-Programming Style) - Section 3

Our group worked together as a team to get this project done. We are proud of the work we did together.  